### ISSUE TITLE
Code folding

### ISSUE THREAD

**Issue by danielzilli** [coteditor/CotEditor#459]
Hello,

Do you have any plan for code folding?

**Comment by 1024jp** [comment]
Currently not.
I don't have fixed plan for it.
I'll consider.

**Comment by roonieone** [comment]
+1

**Comment by danielzilli** [comment]
Any chance?

**Comment by 1024jp** [comment]
@danielzilli 
Still under consideration.
To implement this feature, I need to change (or improve) CotEditor's syntax definition format at first and this is a quite hard task.
Actually, I also want to add this feature, but need time.

**Comment by NSExceptional** [comment]
🙏🏼

**Comment by imkeato** [comment]
+1

**Comment by uschmelzer** [comment]
+1

**Comment by gitgetgot** [comment]
+1

**Comment by fatherofinvention** [comment]
+1 

By the way, love CotEditor. Great work.

**Comment by chrisgrieser** [comment]
Love Coteditor as well. Also want to leave my +1 for the feature.

**Comment by makiato** [comment]
CotEditor is the quick, minimal editor with just enough functions - and this is the only key feature missing for me.

**Comment by BenzLeung** [comment]
+1

**Comment by aidan-gibson** [comment]
Would love this feature as well.

**Comment by georgik-flutter** [comment]
+1

**Comment by theq86** [comment]
+1

**Comment by mayurankv** [comment]
Also a +1 from me!

**Comment by kaizh102** [comment]
+1

**Comment by overfullstack** [comment]
+1

**Comment by 1024jp** [comment]
I privately investigated the possibility of implementation of code folding last years, and found it relates deeply with TextKit framework (and also update of the syntax definition format, I mentioned previously).
Meanwhile, Apple released TextKit 2 at WWDC 2021 and CotEditor needs to migrate to it one day in the future.

The problem is TextKit 2 is actually very different from the current TextKit (1) on which CotEditor depends a lot.
That means it requires a lot of work only to migrate from TextKit 1 to TextKit 2 in the next few years, and I'd like to implement code folding first after the migration is done.
Otherwise, the migration becomes much more complex.

I'm actually interested in implementing code folding even it is hard. I also know a high demand.
But due to the reason above, I have to say it will not be implemented in the next few years, at least.

**Comment by tanranran** [comment]
+1

**Comment by robhailey** [comment]
+1

**Comment by UserNameNinger** [comment]
+1

**Comment by mrichman** [comment]
I'l pile on. +1

**Comment by czanas** [comment]
+1 here as well.

**Comment by wesort** [comment]
+1

**Comment by rpietzsch** [comment]
+1

**Comment by 1024jp** [comment]
Please __don’t just post “+1”__ to existing requests; it makes no difference and adds noise.

cf. [CONTRIBUTING.md](https://github.com/coteditor/CotEditor/blob/main/CONTRIBUTING.md#feature-requests)

**Comment by rpietzsch** [comment]
I agree with our policy, so please accept my apologies for adding _noise_. My perspective on this is: knowing the level of support for a potential feature will give you valuable insight into its relevance. 

Or are you suggesting to do a "👍" on the original post instead of a "+1" in a new response?

**Comment by 1024jp** [comment]
@rpietzsch 

Thank you for the reply.

> Or are you suggesting to do a "👍" on the original post instead of a "+1" in a new response?

Yes, that would be a nice alternative.

**Comment by wehnertb** [comment]
From the previous comments, TextKit 2 was released at WWDC 2021 and we're now in 2025, any sign of potentially getting cold folding any time soon?

It's probably one of the biggest features I would love to see in CotEditor.  It would make it darn near perfect if it had this. :)

**Comment by 1024jp** [comment]
Migrating to TextKit 2 is not an easy task, considering we have approximately 10,000 lines of code related to it. Moreover, the concept of TextKit 2 is quite different from that of TextKit 1.

Nevertheless, I have decided to attempt the migration within the next year. I hope to release the updated version sometime next year. After that, I plan to implement the code folding feature.

**Comment by wehnertb** [comment]
That's great news!  I'm appreciate the feedback and the scale of effort involved.  I really like CotEditor, so far it's one of the best I've found.

**Comment by aaronbuckner** [comment]
Love this editor's simplicity but the code folding is that one thing...
